File: 452/main.py
"""
There are a number of spherical balloons spread in two-dimensional space. For each balloon, provided input is the start and end coordinates of the horizontal diameter. Since it's horizontal, y-coordinates don't matter and hence the x-coordinates of start and end of the diameter suffice. Start is always smaller than end. There will be at most 104 balloons.

An arrow can be shot up exactly vertically from different points along the x-axis. A balloon with xstart and xend bursts by an arrow shot at x if xstart ≤ x ≤ xend. There is no limit to the number of arrows that can be shot. An arrow once shot keeps travelling up infinitely. The problem is to find the minimum number of arrows that must be shot to burst all balloons.

Example:

Input:
[[10,16], [2,8], [1,6], [7,12]]

Output:
2

Explanation:
One way is to shoot one arrow for example at x = 6 (bursting the balloons [2,8] and [1,6]) and another arrow at x = 11 (bursting the other two balloons).
"""

class Solution(object):
    def findMinArrowShots(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        if len(points) == 0:
            return 0
        points = sorted(points, key=lambda x: x[1])
        greedy = points[0]
        end = greedy[1]
        arrows = 1
        for i in range(1, len(points)):
            bal = points[i]
            if bal[0] > end:
                greedy = bal
                arrows += 1
                end = bal[1]
                continue
            else:
                end = min(end, bal[1])
        return arrows

ans = Solution()
print(ans.findMinArrowShots([[1,2],[3,4],[5,6],[7,8]]))
print(ans.findMinArrowShots([[10,16], [2,8], [1,6], [7,12], [1, 3]]))

File: 452/main.ts
function findMinArrowShots(points: number[][]): number {
  let sortedBallons = points.sort((a, b) => b[0] - a[0]);
  let pops = 0;
  while (sortedBallons.length !== 0) {
    let intersection: number[] = [];
    while (true) {
      let balloon = sortedBallons.pop();
      if (!balloon) break;
      if (intersection.length === 0) intersection = balloon;
      let newIntersection = findIntersection(intersection, balloon);
      if (newIntersection.length === 0) {
        sortedBallons.push(balloon);
        break;
      }
      //   console.log(`balloon: ${balloon}`);
      //   console.log(`intersection: ${intersection}`);
      //   console.log(`newIntersection: ${newIntersection}`);
      intersection = newIntersection;
    }
    // console.log("increase pop");
    pops++;
  }
  return pops;
}

function findIntersection(intersection: number[], balloon: number[]): number[] {
  if (intersection[1] < balloon[0] || intersection[0] > balloon[1]) return [];
  return [
    Math.max(intersection[0], balloon[0]),
    Math.min(intersection[1], balloon[1]),
  ];
}

// greedy approach
// from the left
// get intersecting ballons
// ballons that intersect with that intersection
// until we can't intersect
// burst these
// repeat
console.log(
  findMinArrowShots([
    [10, 16],
    [2, 8],
    [1, 6],
    [7, 12],
  ])
);
